### PR TITLE
Consider nested cgroup memory limits for the total memory tracker

### PR DESCRIPTION
**Pull request by rschu1ze**
Review note: Contains commits
- https://github.com/ClickHouse/ClickHouse/pull/59435/commits/92a00d317cddec0e65d398db01ef7ee71fc818a5 - Cosmetics.
- https://github.com/ClickHouse/ClickHouse/pull/59435/commits/ad0c0d87276e7a2a9eedee07a41a55895f7d88a5 - Cosmetics but also fixes the edge case that cgroup memory limit "0" was incorrectly interpreted as "no limit" but it really means "zero" memory.
- https://github.com/ClickHouse/ClickHouse/pull/59435/commits/3243ea0b66ed9b435011aa6570b70ed080cd5410 Kills cgroups v1 supprt for simplicity. V1 has a different hierarchy model which would have made the next commit more complex. Also, v1 is dead since over half a decade.
- https://github.com/ClickHouse/ClickHouse/pull/59435/commits/e997168b50a7d2f471d1d8f552e0b7f6a2df495a See the changelog message
- EDIT: https://github.com/ClickHouse/ClickHouse/pull/59435/commits/80a0e05ac4e3fc6d1837877bad2be7a1016ee4e6 Resurrects cgroups v1. They are still enabled in many Linux distributions in GCP, AWS and Azure :-(
- ad13a57b7f462fa7c0ed0a50a2572969d334e845 Improves the detection of v2 features a bit

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
At startup and whenever config files are changed, ClickHouse updates the hard memory limits of its total memory tracker. These limits are computed based on various server settings and cgroups limits (on Linux). Previously, setting `/sys/fs/cgroup/memory.max` (for cgroups v2) was hard-coded. As a result, cgroup v2 memory limits configured for nested groups (hierarchies), e.g. `/sys/fs/cgroup/my/nested/group/memory.max` were ignored. This is now fixed. The behavior of v1 memory limits remains unchanged.

### CODE REVIEW

File: base/base/getMemoryAmount.cpp [hunk 1]
```diff
@@ -11,6 +13,54 @@
...
+
+    std::string cgroup_name;
+    cgroup_name_file >> cgroup_name;
+
+    /// cgroups v2 will show a single line with prefix "0::/"
+    /// - https://book.hacktricks.xyz/linux-hardening/privilege-escalation/docker-security/cgroups
+    const std::string v2_prefix = "0::/";
+    if (!cgroup_name.starts_with(v2_prefix))
+        return {};
+    cgroup_name = cgroup_name.substr(v2_prefix.length());
```
Comment (nickitat) [inline base/base/getMemoryAmount.cpp:35]: is it guaranteed that there is only a single name?

File: base/base/getMemoryAmount.cpp [hunk 2]
```diff
@@ -26,21 +76,17 @@ uint64_t getMemoryAmountOrZero()
...
+    if (limit.has_value() && *limit < memory_amount)
+         memory_amount = *limit;
     else
     {
-        // CGroups v1
+        /// Cgroups v1 were replaced by v2 in 2015. The only reason we keep supporting v1 is that the transition to v2
+        /// has been slow. Caveat : Hierarchical groups as in v2 are not supported for v1, the location of the memory
+        /// limit (virtual) file is hard-coded.
+        /// TODO: check at the end of 2024 if we can get rid of v1.
         std::ifstream cgroup_limit("/sys/fs/cgroup/memory/memory.limit_in_bytes");
```
Comment (nickitat) [inline base/base/getMemoryAmount.cpp:90]: > also fixes the edge case that cgroup memory limit "0" was incorrectly interpreted as "no limit" but it really means "zero" memory

haven't you resurrected the bug too?)

File: base/base/getMemoryAmount.cpp (hunk 1 above)
Comment (rschu1ze) [reply base/base/getMemoryAmount.cpp:35]: Yes, there is such a guarantee for v2.

It is (un?)fortunately possible to have Linux hosts with mixed v1/v2 usage (see the link to book.hacktricks.xyz), in this case there would be more lines. But afaik, that was only during a short transition period in pre-2020 the case.

Comment (rschu1ze) [reply base/base/getMemoryAmount.cpp:35]: But perhaps I should add a check to make sure the system is v2-only.

File: base/base/getMemoryAmount.cpp (hunk 2 above)
Comment (rschu1ze) [reply base/base/getMemoryAmount.cpp:90]: You are right, fixed, thanks.

File: base/base/getMemoryAmount.cpp [hunk 3]
```diff
@@ -27,14 +27,26 @@ std::optional<uint64_t> getCgroupsV2MemoryLimit()
...
+    /// Make sure that the memory controller is enabled.
+    /// - cgroup.controllers defines which controllers *can* be enabled.
+    /// - cgroup.subtree_control defines which controllers *are* enabled.
+    /// (see https://docs.kernel.org/admin-guide/cgroup-v2.html)
+    /// Caveat: nested groups may disable controllers. For simplicity, check only the top-level group.
+    /// ReadBufferFromFile subtree_control_file(default_cgroups_mount / "cgroup.subtree_control");
+    /// std::string subtree_control;
+    /// readString(subtree_control, subtree_control_file);
+    /// if (subtree_control.find("memory") == std::string::npos)
+    ///     return {};
```
Comment (nickitat) [inline base/base/getMemoryAmount.cpp:39]: why do we keep this code?

File: base/base/getMemoryAmount.cpp [hunk 4]
```diff
@@ -11,6 +14,80 @@
...
```
Comment (nickitat) [inline base/base/getMemoryAmount.cpp:39]: why do we kept this code?

**Review by nickitat — APPROVED** [review]
it is not particularly relevant to this pr, but in general seems like there are quite a few places when any logging might be useful

File: base/base/getMemoryAmount.cpp (hunk 4 above)
Comment (rschu1ze) [reply base/base/getMemoryAmount.cpp:39]: Somehow missed to remove it. I'll delete it in #59516

**Comment by Stanislav-Vlasov** [comment]
I have a problem, I think, related to this change. I'm running the Clickhouse docker image version 24.2.2.71 (official build) on Ubuntu 22.04 (cgroup v2 enabled by default, no cgroup v1) with error in log. This error did not occur when I ran the previous version 24.1 of Clickhouse.

<Error> Application: Disabling cgroup memory observer because of an error during initialization: Code: 107. DB::Exception: Cannot find cgroups v1 or v2 current memory file. (FILE_DOESNT_EXIST), Stack trace (when copying this message, always include the lines below):

0. DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000cf5af1b
1. DB::Exception::Exception<>(int, FormatStringHelperImpl<>) @ 0x00000000078cc583
2. DB::CgroupsMemoryUsageObserver::CgroupsMemoryUsageObserver(std::chrono::duration<long long, std::ratio<1l, 1l>>) @ 0x000000000d11da73
3. DB::Server::main(std::vector<String, std::allocator<String>> const&) @ 0x000000000d0cfedd
4. Poco::Util::Application::run() @ 0x00000000158c9606
5. DB::Server::run() @ 0x000000000d0c0551
6. Poco::Util::ServerApplication::run(int, char**) @ 0x00000000158d2479
7. mainEntryClickHouseServer(int, char**) @ 0x000000000d0bd60a
8. main @ 0x00000000078c8e18
9. ? @ 0x00007f69f6e6d083
10. _start @ 0x00000000060c686e
 (version 24.2.2.71 (official build))

Additional information about my HOST system
1) uname -a
Linux 6.2.0-37-generic 22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Thu Nov  2 18:01:13 UTC 2 x86_64 x86_64 x86_64 GNU/Linux

2) docker -v
Docker version 24.0.2, build cb74dfc

Additional information from the inside of running container
1) cat /proc/1/cgroup
0::/

2) ls /sys/fs/cgroup
cgroup.controllers
cgroup.events
cgroup.freeze
cgroup.kill
cgroup.max.depth
cgroup.max.descendants
cgroup.pressure
cgroup.procs
cgroup.stat
cgroup.subtree_control
cgroup.threads
cgroup.type
cpu.idle
cpu.max
cpu.max.burst
cpu.pressure
cpuset.cpus
cpuset.cpus.effective
cpuset.cpus.partition
cpuset.mems
cpuset.mems.effective
cpu.stat
cpu.uclamp.max
cpu.uclamp.min
cpu.weight
cpu.weight.nice
hugetlb.1GB.current
hugetlb.1GB.events
hugetlb.1GB.events.local
hugetlb.1GB.max
hugetlb.1GB.numa_stat
hugetlb.1GB.rsvd.current
hugetlb.1GB.rsvd.max
hugetlb.2MB.current
hugetlb.2MB.events
hugetlb.2MB.events.local
hugetlb.2MB.max
hugetlb.2MB.numa_stat
hugetlb.2MB.rsvd.current
hugetlb.2MB.rsvd.max
io.bfq.weight
io.max
io.pressure
io.prio.class
io.stat
io.weight
memory.current
memory.events
memory.events.local
memory.high
memory.low
memory.max
memory.min
memory.numa_stat
memory.oom.group
memory.peak
memory.pressure
memory.reclaim
memory.stat
memory.swap.current
memory.swap.events
memory.swap.high
memory.swap.max
memory.zswap.current
memory.zswap.max
misc.current
misc.events
misc.max
pids.current
pids.events
pids.max
pids.peak
rdma.current
rdma.max

**Comment by Stanislav-Vlasov** [comment]
As you can see, there are no nested cgroups in /sys/fs/cgroup

Due to problem of "Disabling cgroup memory observer", the available memory is incorrectly determined. Clickhouse believes that all HOST memory is available without respect of cgroup limit. This causes container memory overruns and triggers OOM killer.

**Comment by rschu1ze** [comment]
@Stanislav-Vlasov First, can you please open a GitHub issue for that? That would be a better place for discussion than the original PR.

As a workaround, you could set a sensible value for server setting `max_server_memory_usage` ([docs](https://clickhouse.com/docs/en/operations/server-configuration-parameters/settings#max_server_memory_usage)). The error just means that ClickHouse will not react to memory limit changes done via cgroups.

Exception `Cannot find cgroups v1 or v2 current memory file` indicates that during database startup, neither a cgroups v1 nor v2 file was found. It would be interesting to debug this, perhaps you have a repro, e.g. the steps you use to invoke docker? 

For completeness, I tested the latest official Docker image (v24.3, Ubuntu 22.04 LTS) and it worked for me, i.e. no errors.

**Pull request merged**